### PR TITLE
Merge 2.9 to 3.1 with sidecar teardown

### PR DESCRIPTION
**Pull request by hpidcock**
Forward ports:
- #15636
- #15650
- #15652
- #15653
- #15633
- #15657
- #15660
- #15661
- #15662
- #15663

Back ports:
- #15336

New changes:
- juju/description/v4 to v4.0.9

Minor conflicts across most files.

### CODE REVIEW

File: api/agent/caasapplication/client.go [hunk 1]
```diff
@@ -42,6 +43,11 @@ func (c *Client) UnitIntroduction(podName string, podUUID string) (*UnitConfig,
 		return nil, err
 	}
 	if err := result.Error; err != nil {
+		if params.IsCodeAlreadyExists(err) {
+			return nil, errors.AlreadyExists
+		} else if params.IsCodeNotAssigned(err) {
+			return nil, errors.NotAssigned
+		}
```
Comment (wallyworld) [inline api/agent/caasapplication/client.go:50]: We could have used `apiservererrors.RestoreError()` here

File: caas/application.go [hunk 2]
```diff
@@ -40,6 +42,9 @@ type Application interface {
 
 	// Units of the application fetched from kubernetes by matching pod labels.
 	Units() ([]Unit, error)
+
+	UnitsToRemove(context.Context, int) ([]string, error)
```
Comment (wallyworld) [inline caas/application.go:46]: Should comment this

**Review by wallyworld — APPROVED** [review]

File: caas/application.go (hunk 2 above)
Comment (hpidcock) [reply caas/application.go:46]: Will handle in 2.9

File: api/agent/caasapplication/client.go (hunk 1 above)
Comment (hpidcock) [reply api/agent/caasapplication/client.go:50]: Will handle in 2.9

**Pull request merged**